bloom-instance/alb.tf

```diff
@@ -0,0 +1,65 @@
+# Anything related to our Application Load Balancer goes here
+
+resource "aws_lb" "public_alb" {
+  name               = "Public-ALB"
```

Comment:
I tried adding the var for the resource prefix but there are limitations specifically for this resource type: prefix can't be longer than 6 characters, and the prefix can only be alphanumeric or hyphens (and our resource prefix is currently "doorway_bloom"). Let me know if you prefer any other prefix than this but we are pretty limited. This seems to be a known issue https://github.com/hashicorp/terraform-provider-aws/issues/1666

Reply:
I meant just make the name to something like `"${var.default_name}-Public-ALB"`, but it looks like that is probably not going to cut it either due to the permitted characters on var.resource_name, the separator characters used for concatenation, and the ridiculously low 32 character limit.  We do still need the name to be unique, so for now go ahead and modify the resource_prefix validation to limit length to 6 characters and change the regex to remove everything but letters, numbers, and hyphens.  Throw a comment in there to note why that var is restricted that way.  Also change the value in tfvars.template to "doorwy" or something like that.

Reply:
Synced offline. Since the name_prefix is really limited for this resource we decided to just use specify the name and to add the workspace and SDLC stage in the name to ensure the ALB name is unique. This means that the workspace name can only be alphanumeric characters or hyphen and must be length 13 at most (given the other characters in the name and knowing that the longest SDLC stage name is 'staging').

Reply:
tl;dr - Let's use "${var.resource_prefix}-${terraform.workspace}-Public-ALB" instead and change our resource_prefix default to "doorway".

The purpose of the resource_prefix var is so we can both visually identify these resources to make them easier to locate in the console and logically separate them to reduce the chances of a naming conflict, whether with this set of templates or any others that might be used.  Our resource names need to provide separation along multiple dimensions: what project it belongs to (scope to a set of templates), what version it is (scope to stage/environment), and the job function it provides in that context (ie "Public-ALB" vs "Private-ALB").  

One thing to consider is that future users of this template may be running it in a single account alongside other projects with their own resources (even if that isn't best practice).  For example, Exygy could theoretically run separate instances with the "dahlia" and "detroit" prefixes.  In that case, both instances would end up trying to create an ALB named "dev-Public-ALB", which would cause the second one to fail.  By adding the resource_prefix first, we can make sure that they both end up with unique names ("dahlia-dev-Public-ALB" and "detroit-dev-Public-ALB", respectively).  Naming doesn't have to be perfect right off the bat, but in some cases changing the name of a resource is a destructive action.  If the fictional Exygy in this case already had "prod-Public-ALB" running by the time they realized the conflict with "dev-Public-ALB", their redeployment to update resource names would take their production system offline for a few minutes (much longer and/or with manual intervention required in some cases) and could potentially have cascading effects.

The problem, that you rightly pointed out previously, is that our current prefix has an underscore, which is not valid for an ALB name.  AWS has a big problem with consistency, especially in naming, and even though it works fine for other resources, we need to change our validation on that var to the lowest common denominator.  We probably don't need to limit the length as mentioned above, but changing the validation to only accept letters, numbers, and hyphens and changing the default value to "doorway" will be sufficient.  At some point in the future we may run into a case where we also have to remove hyphens and stick with alphanumeric characters, but we don't want to overly limit ourselves beyond what is needed just yet.  It's easier to add restrictions later on than it is to remove them.

Reply:
Perfect!